java/arcs/core/allocator/Arc.kt

```diff
@@ -91,56 +104,45 @@ class Arc internal constructor(
-                it(arcState)
-            }
-        }
+    private fun recomputeArcState(states: Collection<ArcState>): ArcState = when {
+        states.any { it == Deleted } -> Deleted
+        states.any { it == Error } -> Error
+        states.all { it == Running } -> Running
+        states.all { it == Stopped } -> Stopped
+        states.all { it == NeverStarted } -> NeverStarted
+        else -> Indeterminate
     }
 
-    private fun recomputeArcState() {
-        sync(this) {
-            val states = arcStatesByHostId.values
-            arcState = when {
-                states.any { it == Deleted } -> Deleted
-                states.any { it == Error } -> Error
-                states.all { it == Running } -> Running
-                states.all { it == Stopped } -> Stopped
-                states.all { it == NeverStarted } -> NeverStarted
-                else -> Indeterminate
-            }
+    private fun maybeRegisterChangeHandlerWithArcHosts() {
+        if (!registered.compareAndSet(false, true)) {
+            return
         }
-    }
-
-    private suspend fun fetchCurrentStates() {
-        partitions.forEach {
-            val arcHost = allocator.lookupArcHost(it.arcHost)
-            arcStatesByHostId[it.arcHost] = arcHost.lookupArcHostStatus(it)
-        }
-    }
 
-    private fun registerChangeHandlerWithArcHosts() = sync(this) {
-        require(registration == null) {
-            "registration called more than once"
-        }
-        runBlocking {
-            // first poll the current states of all hosts
-            fetchCurrentStates()
-
-            // Register event listeners
+        arcStatesByHostFlow = callbackFlow {
             partitions.forEach { partition ->
                 val arcHost = allocator.lookupArcHost(partition.arcHost)
-                registration = arcHost.addOnArcStateChange(id) { _, state ->
-                    sync(this) {
-                        arcStatesByHostId[partition.arcHost] = state
-                        recomputeArcState()
-                    }
+                registrations[partition.arcHost] = arcHost.addOnArcStateChange(id) { _, state ->
+                    offer(partition.arcHost to state)
                 }
             }
+            closeFlow = { close() }
+            awaitClose { unregisterChangeHandlerWithArcHosts() }
+        }.scan(
+            partitions.map { it.arcHost to NeverStarted }.associateBy({ it.first }, { it.second })
+        ) { states, (host, state) ->
+            val newStates = states.toMutableMap()
+            newStates[host] = state
+            newStates
+        }.map {
+            recomputeArcState(it.values)
+        }.onEach { state ->
+            arcState = state
+            arcStateChangeHandlers.value.toList().forEach { handler -> handler(state) }
         }
 
-        recomputeArcState()
+        arcStatesByHostFlow.launchIn(
+            CoroutineScope(coroutineContext + CoroutineName("Arc (flow collector) $id"))
```

Comment:
Suggest `+ Job()` to the coroutineContext unless we ensure no exceptions other than CancellationException happens, otherwise the exception will cancel all its parent jobs.

Reply:
Ah, good catch.